wip for #196 but also fixing #198

petersilva opened the pull request:
close #198 

I have been working on the hpc-mirorring, finding edge cases that don´t work.  One such case is events to reflect directory removals.  I noticed that, as per #198 the events produced by the C  for rmdir events (or rm -rf) were not being produced with the correct fileOp field:

* "fileOp" : { "remove":"" },

when it should have:

* "fileOp" : { "remove":"", "directory":"" }

on the Python side, you need the  both attributes when trying to mirror an rmdir.

So that was one problem.  Another problem was that when doing something like "rm -rf" some of the remove events are for directories.  The way to create a file removal event with the current API is to provide a NULL pointer to the file status struct.  The file status struct normally provides the file type (file, directory link, etc...)
since the removal event doesn't have this information, the API needs to be changed to allow communicating
that the file being removed is a directory.   so added *rmflags* to a whole slew of routines (two or three levels deep.) 

```

   const int rmflags 
   * 0 -- sb MUST not be null, this is not a remove event.
   * 1 -- sb==NULL this is a file (non-directory) removal.
   * 2 -- sb==NULL this is a directory removal.

```

So that's the logic for #198... Unfortunately I was working on that in a branch that was originally about #196.  I tried cherry-picking off the above to just PR cleanly for #198, but it turned out these patches depend on the earlier ones... so I needed to PR the whole thing.

The rest of these patches are cleanups that try to make the code cleaner, while testing thoughts about #196.
None of them fix it. 

My guess as to the real problem is for #196:   when we add process cleanup routine, it gets added after posix thread cleanup has occurred. when we  try to close a file, it calls posix thread code where those data structures have already been cleaned up.

So what are these patches?

* 5fc86ae5882aad0cebefac3a6c7028d4ad727795  segfaults were happenning when calling hash functions, so got rid of some times when calling those routinges is useless.   but since we aren't calling the routines this could change results... but when I ran the flow tests... it seemed fine... 
*  1719b3984a5c602cfe4fd7038c4154d0a541ffa9 - reference through un-initialized pointer... looked wrong.
* 34773816a42d559e6254321c1116fb4cd1ec7d03 -- SSL hash initialization routines... we were using a very old API, (and it was crashing in it.) so I changed it to use a newer API version.  It didn't change anything though.
*  688edcbbbdab8f2bbe55998e21ae3eac65cc8d39 -- SR_SHIMDEBUG log outputs weren't working because any time they were set, it would suppress messages... (a test was inverted.)

petersilva committed:
derive identity for files only.

adding skip of identity generation for hard links.
adding skip of identity generation for non-files.

petersilva committed:
just_the_name was using null pointers

petersilva committed:
there are more files to delete, shim_post_run should expect 6

petersilva committed:
modernized SSL EVP API calls.

still crashes but _create/ replaced by _new & _free (was not cleaning
up before.) It still crashes in init_ex for file removal... odd.

petersilva committed:
polarity of debug levels was wrong.

petersilva committed:
simpler init routine

petersilva committed:
free of a null point not needed

petersilva committed:
local abbreviation for overly complicated variable

petersilva committed:
rmdir processing with api flag instead of global var

there was a global variable rmdir_in_progress, but it wasn't
catching all events. Decided on an API change instead.
now the sr_post related event processing requires an rmflags parameter.

petersilva committed:
correct rmdir event v03 output format

petersilva committed:
Merge branch 'development' into issue196

petersilva commented:
This branch had been started ... oh a month ago... so the conflicts were there based on other merged patches... so I merged to it to get rid of the conflicts.

reidsunderland commented on sr_post.c:
```diff
@@ -1101,9 +1106,9 @@ void sr_post_rename(struct sr_context *sr_c, const char *o, const char *n, const
+				sr_post(sr_c, oldname, &sb, 0);
 			} else {
-				sr_post(sr_c, oldname, NULL);
+				sr_post(sr_c, oldname, NULL, 0);
```
is it correct for `rmflags` to be `0` here, where sb is NULL?

> ```
>    const int rmflags 
>    * 0 -- sb MUST not be null, this is not a remove event.
>    * 1 -- sb==NULL this is a file (non-directory) removal.
>    * 2 -- sb==NULL this is a directory removal.
> ```

reidsunderland reviewed (commented):
The changes look good to me, but I noticed one possible error with the rmflags (I put a comment on it)

petersilva committed:
sr_post() if NULL then last arg cannot be zero

This is an edge case that will likely never happen as it deals
with mirroring using sarra v2 (all mirroring is currently sr3)

found by Reid. but anyways was not obeying current API.

petersilva committed:
increment version with some changelog stuff

petersilva committed:
Revert "derive identity for files only."

This reverts commit 5fc86ae5882aad0cebefac3a6c7028d4ad727795.
